### PR TITLE
Endpoint testing

### LINKED ISSUE #390: Testing Endpoints
As asked [on the Gitter channel](https://gitter.im/finagle/finch?at=55d32e8a4f6c9d3018316d5d), it would be nice to provide a meaning for testing Finch endpoints (routers).

The idea I'm suggesting relies on types we expose from endpoints (that's a very good excuse to have them) and allows the users to treat endpoints as a regular functions so they could be easily tested.

Here is the example.

```scala
case class User(name: String)

val getUser: Endpoint[User] = get("users" / long) { id: Long =>
  User(id.toString)
}

// test
getUser.lift(100) shouldBe User("100")
```

The `lift` (or whatever) method lifts the endpoint to a function `(A, B, ..., Z) => Out`, where `(A, B, ..., Z)` are values extracted fromt the request by endpoint and `Out` is a returned value.

Not sure if it's even possible, but I'd like to move Finch (make required chnages) towards the direction it will be possible.

### PR DESCRIPTION
**Pull request by BenFradet**
fixes #390

### CODE REVIEW

File: core/src/main/scala/io/finch/Input.scala [hunk 1]
```diff
@@ -10,6 +11,21 @@ final case class Input(request: Request, path: Seq[String]) {
   def headOption: Option[String] = path.headOption
   def drop(n: Int): Input = copy(path = path.drop(n))
   def isEmpty: Boolean = path.isEmpty
+
+  def withContent(s: String): Input = {
+    val req = request
+    req.contentString = s
+    req.contentLength = s.length.toLong
+    req.charset = "utf-8"
```
Comment (BenFradet) [inline core/src/main/scala/io/finch/Input.scala]: should we add `charset` as a param to `withContent`?

**Comment by BenFradet** [comment]
I haven't yet found a clean way to do a `withHeaders`.

Regarding `withParams`, shouldn't the user just submit a new `Request` containing those params?

File: core/src/main/scala/io/finch/Input.scala [hunk 2]
```diff
@@ -10,6 +11,21 @@ final case class Input(request: Request, path: Seq[String]) {
   def headOption: Option[String] = path.headOption
   def drop(n: Int): Input = copy(path = path.drop(n))
   def isEmpty: Boolean = path.isEmpty
+
+  def withContent(s: String): Input = {
```
Comment (BenFradet) [inline core/src/main/scala/io/finch/Input.scala]: Should we go with `withContent` or `withBody`?

File: core/src/main/scala/io/finch/Input.scala [hunk 3]
```diff
@@ -19,4 +35,14 @@ object Input {
...
+
+  def get(path: String): Input = Input(Request(Method.Get, path))
+  def post(path: String): Input = Input(Request(Method.Post, path))
+  def put(path: String): Input = Input(Request(Method.Put, path))
+  def patch(path: String): Input = Input(Request(Method.Patch, path))
+  def delete(path: String): Input = Input(Request(Method.Delete, path))
+  def connect(path: String): Input = Input(Request(Method.Connect, path))
+  def head(path: String): Input = Input(Request(Method.Head, path))
+  def options(path: String): Input = Input(Request(Method.Options, path))
+  def trace(path: String): Input = Input(Request(Method.Trace, path))
```
Comment (BenFradet) [inline core/src/main/scala/io/finch/Input.scala]: I don't know if we should support all methods or a restricted set, what do you think?

File: core/src/main/scala/io/finch/Input.scala (hunk 2 above)
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Input.scala]: Let's do `withBody`. You can also make it take `Buf` instead.

File: core/src/main/scala/io/finch/Input.scala (hunk 1 above)
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Input.scala]: Yeah, Let's make it `withBody(buf: Buf, cs: Option[Charset] = None)`.

File: core/src/main/scala/io/finch/Input.scala (hunk 3 above)
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Input.scala]: Let's keep them all, it's fine.

**Comment by vkostyukov** [comment]
Looks pretty great, @BenFradet! Thanks a lot! Mind squashing the commits into one, once you're done with changes?

**Comment by BenFradet** [comment]
@vkostyukov do you have opinions/ideas regarding `withHeaders` and `withParams`?

**Comment by vkostyukov** [comment]
I think `withParams(params: (Stirng, Stirng)*))` and `withHeaders(headers: (String, String)*)` should work fine here.

File: core/src/main/scala/io/finch/Input.scala [hunk 4]
```diff
@@ -10,6 +13,14 @@ final case class Input(request: Request, path: Seq[String]) {
   def headOption: Option[String] = path.headOption
   def drop(n: Int): Input = copy(path = path.drop(n))
   def isEmpty: Boolean = path.isEmpty
+
+  def withBody(buf: Buf, charset: Option[Charset] = None): Input = {
+    val req = request
+    req.content = buf
+    req.contentLength = buf.length.toLong
+    if (charset.isDefined) req.charset = charset.get.displayName()
+    copy(request = req)
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Input.scala]: Since request is mutable we can just return `this`

**Comment by BenFradet** [comment]
@vkostyukov I implemented `withHeaders` but, after much fiddling around with trying to implement `withParams`, I haven't been able to. Do you have ideas on how to do so?

**Comment by vkostyukov** [comment]
Ah right. I see the problem now. Let's think about that.

Looks like we can restrict to the following HTTP verbs for now: GET, PUT, PATCH, DELETE, POST. All of them, except for POST, usually are idempotent, which means users might want to use query string with them. POST is a special case that we should handle carefully and not motivate users to use odd practices/design decisions.

Here is what I think we should do.

* Let's support `Input.$method(path: String, params: (String, String)*)` for GET, PUT, PATCH, DELETE
* Let's support `Input.post(path: String)`
* Let's provide the following methods to specify the payload (on `Input`):
  * `withBody(body: Buf, charset: Option[Charset] = None)` - basic case
  * `withForm(params: (String, String)*)` - URL encoded params/form
  * `withMultipartForm(fileUpload: ???, params: (String, String)*)` - for multipart/form-data payloads

I think this covers enough cases and provides a reasonable API so users can easily test their endpoints. What do you think @BenFradet?

**Comment by BenFradet** [comment]
Fine with me. I still don't see how to implement `withForm`/`withParams` since AFAICT you can't modify the params of an existing `Request`.

**Comment by vkostyukov** [comment]
Yeah, but those go into a payload/body as `application/x-www-form-urlencoded` or `multipart/form-data`. You can take a look at Finagle's `RequestBuilder` on how it's implemented. I think we can either reuse `RequestBuilder` here or call N3 API directly. Ether way works for me.

**Comment by BenFradet** [comment]
Ok, thanks, I'll investigate `RequestBuilder`.

File: core/src/main/scala/io/finch/Input.scala [hunk 5]
```diff
@@ -10,6 +13,31 @@ final case class Input(request: Request, path: Seq[String]) {
...
+    if (params.nonEmpty) {
+      val req = RequestBuilder()
+        .addFormElement(params: _*)
+        .addHeaders(request.headerMap.toMap)
+        .url(request.host.get)
+        .buildFormPost()
+      copy(request = req)
+    } else {
+      this
+    }
```
Comment (BenFradet) [inline core/src/main/scala/io/finch/Input.scala]: I'm seeing a weird NPE from finagle's when `params` is an empty seq (cf [buildFormPost](https://github.com/twitter/finagle/blob/develop/finagle-http/src/main/scala/com/twitter/finagle/http/RequestBuilder.scala#L347-L383).

I guess netty's `HttpPostRequestEncoder` does not like having no body.

File: core/src/main/scala/io/finch/Input.scala [hunk 6]
```diff
@@ -10,6 +13,31 @@ final case class Input(request: Request, path: Seq[String]) {
...
+    headers.foreach { case (k, v) => request.headerMap.set(k, v) }
+    this
+  }
+
+  def withForm(params: (String, String)*): Input = {
+    if (params.nonEmpty) {
+      val req = RequestBuilder()
+        .addFormElement(params: _*)
+        .addHeaders(request.headerMap.toMap)
+        .url(request.host.get)
```
Comment (BenFradet) [inline core/src/main/scala/io/finch/Input.scala]: How would you handle the case when host hasn't been set? Returning the same `Input`?

**Comment by BenFradet** [comment]
@vkostyukov implemented what you advocated for, review welcome

File: core/src/main/scala/io/finch/Input.scala [hunk 7]
```diff
@@ -10,6 +13,40 @@ final case class Input(request: Request, path: Seq[String]) {
...
+  def withMultipartForm(fileElements: FileElement*)(params: (String, String)*): Input = {
+    require(request.host.isDefined, "The host has to be defined")
+    require(fileElements.nonEmpty, "Cannot create request without file elements")
+    val req = RequestBuilder()
+      .addFormElement(params: _*)
+      .add(fileElements)
+      .url(request.host.get)
+      .buildFormPost(multipart = true)
+    copy(request = req)
+  }
```
Comment (BenFradet) [inline core/src/main/scala/io/finch/Input.scala]: I'm not sure `withForm` and `withMultipartForm` should be copy ctor, imo it'd be more appropriate to make them `post` factories, what do you think?

File: core/src/main/scala/io/finch/Input.scala [hunk 8]
```diff
@@ -19,4 +56,26 @@ object Input {
   def apply(req: Request): Input = Input(req, req.path.split("/").toList.drop(1))
 
   implicit val inputEq: Eq[Input] = Eq.fromUniversalEquals
+
+  def get(path: String): Input = Input(Request(path))
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Input.scala]: I don't think we need these methods given that `(String, String)*` defaults to empty list.

File: core/src/main/scala/io/finch/Input.scala [hunk 9]
```diff
@@ -10,6 +13,40 @@ final case class Input(request: Request, path: Seq[String]) {
...
+
+  def withForm(params: (String, String)*): Input = {
+    require(request.host.isDefined, "The host has to be defined")
+    require(params.nonEmpty, "Cannot create request without params")
+    val req = RequestBuilder()
+      .addFormElement(params: _*)
+      .addHeaders(request.headerMap.toMap)
+      .url(request.host.get)
+      .buildFormPost()
+    copy(request = req)
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Input.scala]: Let's just change the request's content, but keep the rest of the things. You can still use `RequestBuilder` to build the request and then copy its body over.

Comment (BenFradet) [inline core/src/main/scala/io/finch/Input.scala]: Will look into that, thx.

Comment (BenFradet) [inline core/src/main/scala/io/finch/Input.scala]: The problem with copying only the body of the request is that headers added by `buildFormPost` won't be preserved.

File: core/src/main/scala/io/finch/Input.scala (hunk 7 above)
Comment (BenFradet) [inline core/src/main/scala/io/finch/Input.scala]: @vkostyukov do you any comment on this?

File: core/src/main/scala/io/finch/Input.scala [hunk 10]
```diff
@@ -10,6 +13,48 @@ final case class Input(request: Request, path: Seq[String]) {
...
+      .url(request.host.get)
+      .buildFormPost()
+    request.method = req.method
+    request.content = req.content
+    if (req.contentLength.isDefined) request.contentLength = req.contentLength.get
+    if (req.charset.isDefined) request.charset = req.charset.get
+    this
+  }
+
+  def withMultipartForm(fileElements: FileElement*)(params: (String, String)*): Input = {
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Input.scala]: How about using `c.t.f.http.exp.Multipart` here? I really don't want to use `FileElement` from the `RequestBuilder` API here, but we already use `Multipart.FileUpload` in Finch so we should be fine.

**Comment by vkostyukov** [comment]
Thanks @BenFradet! This looks pretty great! Just one concern about `FileUpload` type.

File: core/src/main/scala/io/finch/Input.scala (hunk 10 above)
Comment (BenFradet) [inline core/src/main/scala/io/finch/Input.scala]: I'm all for it, that will simplify things. I'll make the necessary changes later today.

**Comment by BenFradet** [comment]
Still haven't figured out how to build a request from a multipart, any help would be appreciated.

File: core/src/main/scala/io/finch/Input.scala [hunk 11]
```diff
@@ -10,6 +13,48 @@ final case class Input(request: Request, path: Seq[String]) {
   def headOption: Option[String] = path.headOption
   def drop(n: Int): Input = copy(path = path.drop(n))
   def isEmpty: Boolean = path.isEmpty
+
+  def withBody(buf: Buf, charset: Option[Charset] = None): Input = {
+    request.content = buf
+    request.contentLength = buf.length.toLong
+    if (charset.isDefined) request.charset = charset.get.displayName()
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Input.scala]: You can do `charset.foreach(cs => request.charset = cs.displayName)`.

File: core/src/main/scala/io/finch/Input.scala (hunk 10 above)
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Input.scala]: I understand `Multipart` is super confusing and I'm not sure I have a good answer for it right now. How about we do everything else and leave multipart for the next attempt?

Comment (BenFradet) [inline core/src/main/scala/io/finch/Input.scala]: I'm ok with that, will remove that later today.

File: core/src/main/scala/io/finch/Input.scala (hunk 11 above)
Comment (BenFradet) [inline core/src/main/scala/io/finch/Input.scala]: Thanks, will do.

File: core/src/test/scala/io/finch/FinchSpec.scala [hunk 12]
```diff
@@ -130,6 +130,13 @@ trait FinchSpec extends FlatSpec with Matchers with Checkers with AllInstances
     )
   } yield b
 
+  def genFileElement: Gen[FileElement] = for {
```
Comment (vkostyukov) [inline core/src/test/scala/io/finch/FinchSpec.scala]: Do we still need those?

Comment (BenFradet) [inline core/src/test/scala/io/finch/FinchSpec.scala]: good point, forgot to remove them

**Comment by BenFradet** [comment]
@vkostyukov good to go :+1:

**Comment by vkostyukov** [comment]
Thanks @BenFradet! I'm going to merge that and wrap up a next release.

**Comment by BenFradet** [comment]
great, looking forward to that!

**Pull request merged**